apps/Gruntfile.js

```diff
@@ -358,98 +359,89 @@ module.exports = function (grunt) {
-    // we don't have to include the large code-studio-common file in the
-    // embedded video page, keeping it fairly lightweight.
-    // (I wonder how much more we could slim it down by removing jQuery!)
-    // @see embed.html.haml
-    {
-      uniqueName: 'embedVideo',
-      entries: {
-        embedVideo: './src/code-studio/embedVideo.js'
-      },
-      provides: ['jquery']
-    },
+          // Build embedVideo.js in its own step (skipping factor-bundle) so that
+          // we don't have to include the large code-studio-common file in the
+          // embedded video page, keeping it fairly lightweight.
+          // (I wonder how much more we could slim it down by removing jQuery!)
+          // @see embed.html.haml
+          embedVideo: './src/code-studio/embedVideo.js',
 
-    // embedBlocks.js is just React, the babel-polyfill, and a few other dependencies
-    // in a bundle to minimize the amound of stuff we need when loading blocks
-    // in an iframe.
-    {
-      uniqueName: 'embedBlocks',
-      entries: {
-        embedBlocks: './src/code-studio/embedBlocks.js'
-      },
-      provides: ['react', 'react-dom', 'radium']
-    },
+          // embedBlocks.js is just React, the babel-polyfill, and a few other dependencies
+          // in a bundle to minimize the amound of stuff we need when loading blocks
+          // in an iframe.
+          embedBlocks: './src/code-studio/embedBlocks.js',
 
-    {
-      uniqueName: 'makerlabDependencies',
-      entries: {
-        makerlab: './src/code-studio/makerlab/makerlabDependencies.js'
-      },
-      provides: ['johnny-five', 'playground-io', 'chrome-serialport']
-    },
+          makerlab: './src/code-studio/makerlab/makerlabDependencies.js',
 
-    {
-      uniqueName: 'pd',
-      entries: {
-        pd: './src/code-studio/pd/workshop_dashboard/workshop_dashboard.jsx'
-      }
-    },
+          pd: './src/code-studio/pd/workshop_dashboard/workshop_dashboard.jsx',
 
-    {
-      uniqueName: 'publicKeyCryptography',
-      entries: {
-        publicKeyCryptography: './src/publicKeyCryptography/main.js'
-      },
-      provides: ['react', 'react-dom', 'radium']
+          publicKeyCryptography: './src/publicKeyCryptography/main.js',
+        }
+      ),
+      externals: [
+        {
+          'jquery': 'var $',
+        }
+      ],
+      plugins: [
+        new webpack.optimize.CommonsChunkPlugin({
+          name: 'common',
+          chunks: _.keys(appsEntries),
+          minChunks: 2
+        }),
+        new webpack.optimize.CommonsChunkPlugin({
+          name: 'code-studio-common',
+          chunks: _.keys(codeStudioEntries).concat(['common']),
+          minChunks: 2
+        }),
+        new webpack.optimize.CommonsChunkPlugin({
+          name: 'minimal',
+          minChunks: 2,
+          chunks: [
+            'plc',
+            'pd',
+            'code-studio-common',
+          ]
+        }),
```

Comment:
Wondering what these CommonsChunkPlugins are doing? This is roughly how it works:

- We look at all the modules that are used by at least two different "apps" related entries (i.e. applab.js and gamelab.js) and pull all the shared modules into a `common.js` file. So any time you add a script tag for `applab.js`, you must add one for `common.js` first.

- Next we look at all the modules that are used by the various "code-studio" entries, along with those ones that got pulled into the `common.js` chunk. Any module that gets used in more than one entry is extracted into `code-studio-common.js`. So now `common.js` depends on `code-studio-common.js` having been loaded.

- Finally, we look at the `plc` and `pd` entry points and grab all the modules shared by them and `code-studio-common.js` and move them into a `minimal.js` file.

The net result is the following:
- for `pd.js` and `plc.js`, you need to also include `minimal.js` for them to work.
- for `levelbuilder.js` et. al. you need to include both `minimal.js` and `code-studio-common.js` (fortunately, `minimal.js` will probably have been cached at this point)
- for `applab.js` you need to include `minimal.js`, `code-studio-common.js` and `common.js`.

While there is some overhead to downloading multiple files like this, I think the file size savings make up for it. I've thought about how to verify that, but I'm afraid that will be a lot of work since it requires production logging that takes into account user behavior (what order of pages people tend to visit and what the browser cache hit rate looks like). Anyway, I don't think it is a big enough problem to stop us from going forward with this change.